### DISCUSSION TITLE
Force error event within an eventListener

### DISCUSSION THREAD

**Discussion by samlising** [solspace/craft-freeform#675] [Questions & Answers] [answered]
I am using freeform 4 and accessing an external api within the "freeform-ajax-before-submit" listener. 

If the external api returns and error I would like to display the error in an error banner and abort the submit process. 

I am able to use event.preventDefault(); to stop the submit process, however the submit button gets stuck on "loading..." 

Is there a call to elegantly cancel the submit process.

This is an example of what I am using so far:

```
formElement.addEventListener('freeform-ajax-before-submit', function (event) {
	var f = event.target.freeform;
	var d = event.data;

	createTicket(data)
	.done( function (result, status, xhr) {
		d.set("ticketId", result.id);
	})
	.fail(function (result, status){
		f._renderFormErrors("ERROR");
		event.preventDefault();
	})
});
```

**Comment by gustavs-gutmanis** [comment, marked as answer]
Hey @samlising!

The issue you're having is due to the `createTicket()` call being async, and working with a Promise.
While the ticket is being created, the event has already finished its call and continues onwards, and only after the ticket has finished being created, the `event.preventDefault()` gets called.

Could you try making the listener into an `async` function and await the result instead?

```js
formElement.addEventListener(
  'freeform-ajax-before-submit', 
  async function (event) {
    var f = event.target.freeform;
    var d = event.data;

    try {
      const result = await createTicket(data);
      d.set("ticketId", result.id);
    } catch (err) {
      f._renderFormErrors(["ERROR"]);
      event.preventDefault();
      
      // To make sure the loading spinner gets removed and the button gets re-enabled, call this:
      f.unlockSubmit();
    }
  }
);
```

You might need to adjust the `try/catch` block to match your case better, but hopefully this takes you in the right direction.
